File: src/airport/graph.cpp
// graph.cpp -- the given guidance-line graph container.
#include "planning/airport/graph.hpp"

#include <algorithm>

namespace planning::airport {

const char* toString(GateKind kind) {
  switch (kind) {
    case GateKind::None:
      return "none";
    case GateKind::Soft:
      return "soft";
    case GateKind::Hard:
      return "hard";
  }
  return "none";
}

VertexId TaxiGraph::addVertex(const Vec2& p, std::string name) {
  Vertex v;
  v.id = static_cast<VertexId>(vertices_.size());
  v.p = p;
  v.name = std::move(name);
  vertices_.push_back(std::move(v));
  return vertices_.back().id;
}

EdgeId TaxiGraph::addEdge(VertexId from, VertexId to, std::string taxiway) {
  Edge e;
  e.id = static_cast<EdgeId>(edges_.size());
  e.from = from;
  e.to = to;
  e.taxiway = std::move(taxiway);
  edges_.push_back(std::move(e));
  return edges_.back().id;
}

VertexId TaxiGraph::findVertex(const std::string& name) const {
  for (const Vertex& v : vertices_)
    if (v.name == name) return v.id;
  return kNoVertex;
}

VertexId TaxiGraph::vertexAt(const Vec2& p, double tol) const {
  for (const Vertex& v : vertices_)
    if (distance(v.p, p) <= tol) return v.id;
  return kNoVertex;
}

Segment TaxiGraph::geometry(EdgeId id) const {
  const Edge& e = edge(id);
  return Segment{vertex(e.from).p, vertex(e.to).p};
}

double TaxiGraph::length(EdgeId id) const { return geometry(id).length(); }

VertexId TaxiGraph::tail(const DirectedEdge& d) const {
  const Edge& e = edge(d.edge);
  return d.forward ? e.from : e.to;
}

VertexId TaxiGraph::head(const DirectedEdge& d) const {
  const Edge& e = edge(d.edge);
  return d.forward ? e.to : e.from;
}

Vec2 TaxiGraph::tailPoint(const DirectedEdge& d) const { return vertex(tail(d)).p; }
Vec2 TaxiGraph::headPoint(const DirectedEdge& d) const { return vertex(head(d)).p; }

double TaxiGraph::heading(const DirectedEdge& d) const {
  return angleOf(headPoint(d) - tailPoint(d));
}

Pose TaxiGraph::poseAlong(const DirectedEdge& d, double s) const {
  const Vec2 a = tailPoint(d);
  const Vec2 b = headPoint(d);
  const double len = distance(a, b);
  const double t = len < kEps ? 0.0 : std::clamp(s, 0.0, len) / len;
  return Pose{a + (b - a) * t, angleOf(b - a)};
}

double TaxiGraph::projectOnto(const DirectedEdge& d, const Vec2& q) const {
  const Vec2 a = tailPoint(d);
  const Vec2 b = headPoint(d);
  const Vec2 ab = b - a;
  const double dd = dot(ab, ab);
  if (dd < kEps) return 0.0;
  return std::clamp(dot(q - a, ab) / dd, 0.0, 1.0) * std::sqrt(dd);
}

const std::vector<EdgeId>& TaxiGraph::incident(VertexId v) const {
  static const std::vector<EdgeId> kEmpty;
  const std::size_t i = static_cast<std::size_t>(v);
  return i < incident_.size() ? incident_[i] : kEmpty;
}

std::vector<DirectedEdge> TaxiGraph::leaving(VertexId v) const {
  std::vector<DirectedEdge> out;
  for (EdgeId e : incident(v)) {
    const Edge& edgeRef = edge(e);
    if (edgeRef.from == v) out.push_back({e, true});
    if (edgeRef.to == v && !edgeRef.oneWay) out.push_back({e, false});
  }
  return out;
}

std::vector<DirectedEdge> TaxiGraph::arriving(VertexId v) const {
  std::vector<DirectedEdge> out;
  for (EdgeId e : incident(v)) {
    const Edge& edgeRef = edge(e);
    if (edgeRef.to == v) out.push_back({e, true});
    if (edgeRef.from == v && !edgeRef.oneWay) out.push_back({e, false});
  }
  return out;
}

void TaxiGraph::build() {
  incident_.assign(vertices_.size(), {});
  for (const Edge& e : edges_) {
    // .at(), so that an edge pointing at a vertex that does not exist is a
    // reportable error rather than a crash in somebody else's test.
    incident_.at(static_cast<std::size_t>(e.from)).push_back(e.id);
    if (e.to != e.from) incident_.at(static_cast<std::size_t>(e.to)).push_back(e.id);
  }
}

}  // namespace planning::airport

File: src/airport/graph.ts
// The guidance-line graph, and the directed-edge state space every on-graph
// search in the capstone runs over.
//
// Part of the *given* library.  The container is given; splitting the raw graph
// at zone boundaries (Step 2) is Exercise 02, and filtering it (Step 5) is
// Exercise 05.

import {
  add,
  angleOf,
  distance,
  dot,
  EPS,
  type Pose,
  scale,
  type Segment,
  sub,
  type Vec2,
} from './geometry.js';
import type { ZoneClass } from './zones.js';

export type VertexId = number;
export type EdgeId = number;

export const NO_VERTEX: VertexId = -1;
export const NO_EDGE: EdgeId = -1;

/**
 * What crossing this vertex costs you.
 *
 *   none  an ordinary junction or a mid-line split
 *   soft  passable, but it generates a communication event and possibly a stop
 *         (apron/taxiway, the movement/non-movement boundary, stand entry)
 *   hard  impassable unless the clearance names this crossing
 */
export type GateKind = 'none' | 'soft' | 'hard';

export const GATE_KINDS: readonly GateKind[] = ['none', 'soft', 'hard'];

export interface Vertex {
  id: VertexId;
  p: Vec2;
  name: string;
  gate: GateKind;
  /** For a hard gate: the runway identifiers on the far side, e.g. ["18", "36"]. */
  protects: string[];
  /**
   * The hold-short line this gate sits on, or -1.  A hard gate without one is a
   * runway edge rather than a painted holding position.
   */
  holdShortId: number;
  /** What the gate separates, filled in by Exercise 02. */
  innerZone: ZoneClass;
  outerZone: ZoneClass;
}

export interface Edge {
  id: EdgeId;
  from: VertexId;
  to: VertexId;
  /** "A", "B", "RWY 09/27", "STAND 2", "APRON", "DEICE" */
  taxiway: string;
  zone: ZoneClass;
  hotspot: boolean;
  closed: boolean;
  /** When true, only from -> to may be travelled. */
  oneWay: boolean;
  /** metres */
  maxWingspan: number;
  maxWeightTonnes: number;
}

/**
 * A state of the on-graph search: an edge *and* the direction it is being
 * travelled, so that heading is part of the state and a turn at a junction can
 * be checked against the minimum turn radius.  Compare Part 1, where the state
 * was just a point.
 */
export interface DirectedEdge {
  readonly edge: EdgeId;
  readonly forward: boolean;
}

export const NO_DIRECTED_EDGE: DirectedEdge = { edge: NO_EDGE, forward: true };

export function directed(edge: EdgeId, forward = true): DirectedEdge {
  return { edge, forward };
}

export function isValidEdge(d: DirectedEdge): boolean {
  return d.edge !== NO_EDGE;
}

/** A dense index, so cost-to-go can live in a flat array. */
export function directedIndex(d: DirectedEdge): number {
  return d.edge * 2 + (d.forward ? 0 : 1);
}

export function reversed(d: DirectedEdge): DirectedEdge {
  return { edge: d.edge, forward: !d.forward };
}

export function sameDirectedEdge(a: DirectedEdge, b: DirectedEdge): boolean {
  return a.edge === b.edge && a.forward === b.forward;
}

export class TaxiGraph {
  private readonly vertexList: Vertex[] = [];
  private readonly edgeList: Edge[] = [];
  private incidentLists: EdgeId[][] = [];

  addVertex(p: Vec2, name: string): VertexId {
    const v: Vertex = {
      id: this.vertexList.length,
      p,
      name,
      gate: 'none',
      protects: [],
      holdShortId: -1,
      innerZone: 'unknown',
      outerZone: 'unknown',
    };
    this.vertexList.push(v);
    return v.id;
  }

  addEdge(from: VertexId, to: VertexId, taxiway: string): EdgeId {
    const e: Edge = {
      id: this.edgeList.length,
      from,
      to,
      taxiway,
      zone: 'unknown',
      hotspot: false,
      closed: false,
      oneWay: false,
      maxWingspan: 80.0,
      maxWeightTonnes: 600.0,
    };
    this.edgeList.push(e);
    return e.id;
  }

  get numVertices(): number {
    return this.vertexList.length;
  }

  get numEdges(): number {
    return this.edgeList.length;
  }

  get numDirectedEdges(): number {
    return 2 * this.numEdges;
  }

  get vertices(): readonly Vertex[] {
    return this.vertexList;
  }

  get edges(): readonly Edge[] {
    return this.edgeList;
  }

  /**
   * Bounds checked on purpose.  A half-written exercise hands these a NO_VERTEX
   * sooner or later, and an exception naming the bad index beats `undefined`
   * spreading quietly through the arithmetic until a route comes out as NaN.
   */
  vertex(v: VertexId): Vertex {
    const found = this.vertexList[v];
    if (!found) throw new RangeError(`no vertex ${v} on this graph`);
    return found;
  }

  edge(e: EdgeId): Edge {
    const found = this.edgeList[e];
    if (!found) throw new RangeError(`no edge ${e} on this graph`);
    return found;
  }

  findVertex(name: string): VertexId {
    for (const v of this.vertexList) if (v.name === name) return v.id;
    return NO_VERTEX;
  }

  /** The vertex within `tol` of p, or NO_VERTEX. */
  vertexAt(p: Vec2, tol = 1e-6): VertexId {
    for (const v of this.vertexList) if (distance(v.p, p) <= tol) return v.id;
    return NO_VERTEX;
  }

  geometry(e: EdgeId): Segment {
    const found = this.edge(e);
    return { a: this.vertex(found.from).p, b: this.vertex(found.to).p };
  }

  length(e: EdgeId): number {
    const g = this.geometry(e);
    return distance(g.a, g.b);
  }

  tail(d: DirectedEdge): VertexId {
    const e = this.edge(d.edge);
    return d.forward ? e.from : e.to;
  }

  head(d: DirectedEdge): VertexId {
    const e = this.edge(d.edge);
    return d.forward ? e.to : e.from;
  }

  tailPoint(d: DirectedEdge): Vec2 {
    return this.vertex(this.tail(d)).p;
  }

  headPoint(d: DirectedEdge): Vec2 {
    return this.vertex(this.head(d)).p;
  }

  heading(d: DirectedEdge): number {
    return angleOf(sub(this.headPoint(d), this.tailPoint(d)));
  }

  /** The pose at arclength s from the tail. */
  poseAlong(d: DirectedEdge, s: number): Pose {
    const a = this.tailPoint(d);
    const b = this.headPoint(d);
    const len = distance(a, b);
    const t = len < EPS ? 0 : Math.min(Math.max(s, 0), len) / len;
    return { p: add(a, scale(sub(b, a), t)), heading: angleOf(sub(b, a)) };
  }

  /** Arclength of the closest point on the edge to q, clamped to the edge. */
  projectOnto(d: DirectedEdge, q: Vec2): number {
    const a = this.tailPoint(d);
    const b = this.headPoint(d);
    const ab = sub(b, a);
    const dd = dot(ab, ab);
    if (dd < EPS) return 0;
    const t = Math.min(Math.max(dot(sub(q, a), ab) / dd, 0), 1);
    return t * Math.sqrt(dd);
  }

  incident(v: VertexId): readonly EdgeId[] {
    return this.incidentLists[v] ?? [];
  }

  /** Every directed edge that leaves v, honouring the one-way flag. */
  leaving(v: VertexId): DirectedEdge[] {
    const out: DirectedEdge[] = [];
    for (const e of this.incident(v)) {
      const found = this.edge(e);
      if (found.from === v) out.push({ edge: e, forward: true });
      if (found.to === v && !found.oneWay) out.push({ edge: e, forward: false });
    }
    return out;
  }

  /** Every directed edge that arrives at v, honouring the one-way flag. */
  arriving(v: VertexId): DirectedEdge[] {
    const out: DirectedEdge[] = [];
    for (const e of this.incident(v)) {
      const found = this.edge(e);
      if (found.to === v) out.push({ edge: e, forward: true });
      if (found.from === v && !found.oneWay) out.push({ edge: e, forward: false });
    }
    return out;
  }

  /** Rebuilds the incidence lists.  Call after a batch of addEdge() calls. */
  build(): void {
    this.incidentLists = this.vertexList.map(() => [] as EdgeId[]);
    for (const e of this.edgeList) {
      const from = this.incidentLists[e.from];
      if (!from) throw new RangeError(`edge ${e.id} starts at vertex ${e.from}, which does not exist`);
      from.push(e.id);
      if (e.to !== e.from) {
        const to = this.incidentLists[e.to];
        if (!to) throw new RangeError(`edge ${e.id} ends at vertex ${e.to}, which does not exist`);
        to.push(e.id);
      }
    }
  }
}
